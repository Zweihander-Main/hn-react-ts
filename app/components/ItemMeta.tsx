import React from 'react';
import { Link } from 'react-router-dom';
import { formatDate } from '../utils/helpers';
import { ThemeConsumer } from '../contexts/theme';

interface ItemMetaProps {
	by: string;
	time: number;
	id: number;
	descendants?: number;
	item?: HNItem;
}

const PostMetaInfo: React.FC<ItemMetaProps> = ({
	by,
	time,
	id,
	descendants,
	item,
}: ItemMetaProps): JSX.Element => {
	return (
		<ThemeConsumer>
			{({ theme }: AppState): JSX.Element => (
				<div className={`meta-info-${theme}`}>
					<span>
						by <Link to={`/user?id=${by}`}>{by}</Link>
					</span>
					<span>on {formatDate(time)}</span>
					{typeof descendants === 'number' && (
						<span>
							with{' '}
							<Link
								to={{
									pathname: '/post',
									search: `?id=${id}`,
									state: {
										item: item,
									},
								}}
							>
								{descendants}
							</Link>{' '}
							comments [
							<a
								href={`https://news.ycombinator.com/item?id=${id}`}
							>
								hn
							</a>
							]
						</span>
					)}
				</div>
			)}
		</ThemeConsumer>
	);
};

export default PostMetaInfo;
