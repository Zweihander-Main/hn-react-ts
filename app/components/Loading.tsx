import * as React from 'react';
import * as PropTypes from 'prop-types';

const styles = {
	content: {
		fontSize: '35px',
		marginTop: '20px',
		textAlign: 'center' as 'center',
	},
};

interface LoadingProps extends React.Props<Loading> {
	text?: string;
	speed?: number;
}

interface LoadingState {
	content: string;
}

export default class Loading extends React.Component<
	LoadingProps,
	LoadingState
> {
	static propTypes = {
		text: PropTypes.string.isRequired,
		speed: PropTypes.number.isRequired,
	};

	static defaultProps = {
		text: 'Loading',
		speed: 300,
	};

	interval: number;
	state = {
		content: this.props.text,
	};

	componentDidMount(): void {
		const { speed, text } = this.props;

		this.interval = window.setInterval((): void => {
			this.state.content === text + '...'
				? this.setState({ content: text })
				: this.setState(
						({ content }: LoadingState): LoadingState => ({
							content: content + '.',
						})
				  );
		}, speed);
	}

	componentWillUnmount(): void {
		window.clearInterval(this.interval);
	}

	render(): JSX.Element {
		return <p style={styles.content}>{this.state.content}</p>;
	}
}
