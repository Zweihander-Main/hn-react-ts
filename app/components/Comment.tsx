import * as React from 'react';
import ItemMeta from './ItemMeta';

interface CommentProps {
	comment: HNItem;
	depth: number;
	toggleCollapse: () => void;
	collapsed: boolean;
}

const Comment: React.FC<CommentProps> = ({
	comment,
	depth,
	toggleCollapse,
	collapsed,
}: CommentProps): JSX.Element => {
	const { id, by, time, text } = comment;

	return (
		<div style={{ marginLeft: 40 * depth }} className="comment">
			<a className="toggleCollapse" href="#" onClick={toggleCollapse}>
				[{collapsed === true ? '+' : '-'}]
			</a>
			{collapsed !== true && (
				<React.Fragment>
					<ItemMeta by={by} time={time} id={id} />
					<p
						dangerouslySetInnerHTML={{
							__html: text,
						}}
					/>
				</React.Fragment>
			)}
		</div>
	);
};

export default Comment;
