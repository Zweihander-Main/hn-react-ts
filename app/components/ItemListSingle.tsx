import * as React from 'react';
import { HNItem as propTypesHNItem } from '../globals.PropTypes';
import ItemMeta from './ItemMeta';

interface ItemListSingleProps {
	item: HNItem;
}

/**
 * Renders a single post in a list of posts
 *
 * @class      ItemListSingle
 * @return     {JSX.Element}
 */
const ItemListSingle: React.FC<ItemListSingleProps> = ({
	item,
}: ItemListSingleProps): JSX.Element => {
	const { title, url, by, time, id, descendants } = item;

	return (
		<li className="post">
			<a className="link" href={url}>
				{title}
			</a>
			<ItemMeta
				by={by}
				time={time}
				descendants={descendants}
				id={id}
				item={item}
			/>
		</li>
	);
};

ItemListSingle.propTypes = {
	item: propTypesHNItem.isRequired,
};

export default ItemListSingle;
