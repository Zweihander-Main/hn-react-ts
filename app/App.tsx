import * as React from 'react';
import { ThemeProvider } from './contexts/theme';
import { BrowserRouter as Router, Route, Switch } from 'react-router-dom';
import { Helmet } from 'react-helmet';
import './index.css';
import Nav from './components/Nav';
import Loading from './components/Loading';
import { AppState } from './types';

const ItemsPage = React.lazy(() => import('./components/ItemsPage'));
const PostPage = React.lazy(() => import('./components/PostPage'));
const UserPage = React.lazy(() => import('./components/UserPage'));

/**
 * Sets up application including meta information, theme context, and routing
 */
const App: React.FC = () => {
	const [theme, setTheme] = React.useState<AppState>('light');

	const toggleTheme = (): void => {
		setTheme((theme) => (theme === 'light' ? 'dark' : 'light'));
	};

	return (
		<Router>
			<Helmet defaultTitle="HackerNews" titleTemplate="HN - %s" />
			<ThemeProvider value={theme}>
				<div className={theme}>
					<div className="container">
						<Nav toggleTheme={toggleTheme} />
						<React.Suspense fallback={<Loading />}>
							<Switch>
								<Route exact path="/">
									<ItemsPage type="top" />
								</Route>
								<Route path="/new">
									<ItemsPage type="new" />
								</Route>
								<Route path="/best">
									<ItemsPage type="best" />
								</Route>
								<Route path="/ask">
									<ItemsPage type="ask" />
								</Route>
								<Route path="/show">
									<ItemsPage type="show" />
								</Route>
								<Route path="/job">
									<ItemsPage type="job" />
								</Route>
								<Route path="/post" component={PostPage} />
								<Route path="/user" component={UserPage} />
								<Route>
									<h1>404</h1>
								</Route>
							</Switch>
						</React.Suspense>
					</div>
				</div>
			</ThemeProvider>
		</Router>
	);
};

export default App;
