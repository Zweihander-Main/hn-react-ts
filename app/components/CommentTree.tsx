import * as React from 'react';
import * as PropTypes from 'prop-types';
import { HNItem as propTypesHNItem } from '../globals.PropTypes';
import { fetchComments } from '../utils/api';
import Loading from './Loading';
import Comment from './Comment';
import InfiniteScroll from 'react-infinite-scroll-component';

interface CommentTreeProps extends React.Props<CommentTree> {
	parent: HNItem;
	depth: number;
}

interface CommentTreeState {
	comments: Array<HNItem>;
	loading: boolean;
	hasMore: boolean;
	loadedComments: Array<HNItem>;
	error: string;
	collapsed: boolean;
}

const commentsToLoad = 5;

export default class CommentTree extends React.Component<
	CommentTreeProps,
	CommentTreeState
> {
	static defaultProps = {
		depth: -1,
	};

	static propTypes = {
		parent: propTypesHNItem.isRequired,
		depth: PropTypes.number.isRequired,
	};

	state = {
		comments: [],
		loading: true,
		hasMore: false,
		loadedComments: [],
		error: null,
		collapsed: false,
	};

	componentDidMount(): void {
		const { parent } = this.props;

		fetchComments(parent.kids || [])
			.then((comments: Array<HNItem>): void =>
				this.setState({
					comments,
					loading: false,
					loadedComments: comments.slice(0, commentsToLoad),
					hasMore: comments.length > 0 ? true : false,
				})
			)
			.catch(({ message }: { message: string }): void =>
				this.setState({
					error: message,
					loading: false,
					hasMore: false,
				})
			);
	}

	addMorePosts = (): void => {
		this.setState(
			(prevState: CommentTreeState): CommentTreeState => {
				const {
					loadedComments,
					comments,
					loading,
					error,
					collapsed,
				} = prevState;
				const newLength = loadedComments.length + commentsToLoad;
				const newComments = comments.slice(0, newLength);

				return {
					loadedComments: newComments,
					hasMore: comments.length > newLength ? true : false,
					comments,
					loading,
					error,
					collapsed,
				};
			}
		);
	};

	toggleCollapse = (e: React.MouseEvent<HTMLElement>): void => {
		e.preventDefault();
		this.setState((prevState) => ({
			...prevState,
			collapsed: !prevState.collapsed,
		}));
	};

	render(): JSX.Element {
		const {
			loadedComments,
			loading,
			hasMore,
			error,
			collapsed,
		} = this.state;
		const { parent, depth } = this.props;

		if (error) {
			return <p className="center-text error">{error}</p>;
		}

		return (
			<React.Fragment>
				{loading === true ? (
					<Loading text="Loading comment" />
				) : (
					<React.Fragment>
						{depth !== -1 && (
							<Comment
								comment={parent}
								depth={depth}
								toggleCollapse={this.toggleCollapse}
								collapsed={collapsed}
							/>
						)}
						{collapsed !== true && (
							<InfiniteScroll
								dataLength={loadedComments.length}
								next={this.addMorePosts}
								hasMore={hasMore}
								loader={
									<Loading text="Loading more comments" />
								}
							>
								{loadedComments.map((comment: HNItem) => (
									<CommentTree
										key={comment.id}
										parent={comment}
										depth={depth + 1}
									/>
								))}
							</InfiniteScroll>
						)}
					</React.Fragment>
				)}
			</React.Fragment>
		);
	}
}
