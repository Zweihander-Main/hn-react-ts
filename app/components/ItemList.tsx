import * as React from 'react';
import * as PropTypes from 'prop-types';
import ItemListSingle from './ItemListSingle';
import InfiniteScroll from 'react-infinite-scroll-component';
import Loading from './Loading';
import { HNItemPT as propTypesHNItem, HNItem } from '../types';

interface ItemListProps {
	items?: Array<HNItem>;
}

interface ItemListReducerState {
	hasMore: boolean;
	loadedItems: Array<HNItem>;
}

const ITEMS_TO_LOAD = 50;

type ItemListReducerActions =
	| { type: '' }
	| { type: 'addMorePosts'; items: Array<HNItem> };

function itemListReducer(
	state: ItemListReducerState,
	action: ItemListReducerActions
): ItemListReducerState {
	switch (action.type) {
		case '': {
			return { ...state };
		}
		case 'addMorePosts': {
			const { loadedItems } = state;
			const { items } = action;
			const newLength = loadedItems.length + ITEMS_TO_LOAD;
			const newItems = items.slice(0, newLength);

			return {
				...state,
				loadedItems: newItems,
				hasMore: items.length > newLength ? true : false,
			};
		}
	}
}

/**
 * Creates list of posts from given array of post data. Will lazy load/infinite
 * scroll data beyond itemsToLoad limit.
 */
const ItemList: React.FC<ItemListProps> = ({ items }: ItemListProps) => {
	const [state, dispatch] = React.useReducer(
		itemListReducer,
		items?.length > 0
			? {
					hasMore: true,
					loadedItems: items.slice(0, ITEMS_TO_LOAD),
			  }
			: {
					hasMore: false,
					loadedItems: [],
			  }
	);

	return (
		<ul>
			{state.loadedItems?.length > 0 ? (
				<InfiniteScroll
					dataLength={state.loadedItems.length}
					next={() => dispatch({ type: 'addMorePosts', items })}
					hasMore={state.hasMore}
					loader={<Loading text="Loading more posts" />}
				>
					{state.loadedItems.map((item: HNItem) => (
						<ItemListSingle key={item.id} item={item} />
					))}
				</InfiniteScroll>
			) : (
				<p className="center-text">No items to display</p>
			)}
		</ul>
	);
};

ItemList.propTypes = {
	items: PropTypes.arrayOf(propTypesHNItem),
};

export default ItemList;
