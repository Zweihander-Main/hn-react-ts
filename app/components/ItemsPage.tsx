import * as React from 'react';
import ItemList from './ItemList';
import Loading from './Loading';
import { fetchMainPosts } from '../utils/api';
import Helmet from 'react-helmet';

interface FetchItemsProps extends React.Props<FetchItems> {
	type: HNTypes;
}

interface FetchItemsState {
	items: Array<HNItem>;
	error: string;
	loading: boolean;
}

export default class FetchItems extends React.Component<
	FetchItemsProps,
	Readonly<FetchItemsState>
> {
	state = {
		items: null,
		error: null,
		loading: true,
	};

	componentDidMount(): void {
		this.handleFetch();
	}

	componentDidUpdate(prevProps): void {
		if (prevProps.type !== this.props.type) {
			this.handleFetch();
		}
	}

	handleFetch = (): void => {
		this.setState({
			items: null,
			error: null,
			loading: true,
		});

		const { type } = this.props;

		fetchMainPosts(type)
			.then((items: Array<HNItem>): void => {
				this.setState({ items, error: null, loading: false });
			})
			.catch(({ message }: { message: string }): void =>
				this.setState({
					error: message,
					loading: false,
				})
			);
	};

	render(): JSX.Element | JSX.Element[] {
		const { items, error, loading } = this.state;
		const { type } = this.props;

		const title = `${type[0].toUpperCase() +
			type.slice(1).toLowerCase()} Stories`;
		const titleJSX = (
			<Helmet key={`title-${title}`}>
				<title>{title}</title>
			</Helmet>
		);

		if (loading === true) {
			return [titleJSX, <Loading key={`loading-${title}`} />];
		}

		if (error) {
			return [
				titleJSX,
				<p key={`error-${title}`} className="center-text error">
					{error}
				</p>,
			];
		}

		return [titleJSX, <ItemList key={`items-${title}`} items={items} />];
	}
}
