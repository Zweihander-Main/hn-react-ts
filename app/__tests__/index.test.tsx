import React from 'react';
import { BrowserRouter as Router } from 'react-router-dom';
import App from '../App';
import Comment from '../components/Comment';
import CommentTree from '../components/CommentTree';
import ItemList from '../components/ItemList';
import ItemListSingle from '../components/ItemListSingle';
import ItemMeta from '../components/ItemMeta';
import ItemsPage from '../components/ItemsPage';
import Loading from '../components/Loading';
import Nav from '../components/Nav';
import PostPage from '../components/PostPage';
import UserPage from '../components/UserPage';

import { render } from '@testing-library/react';

const testComment = { id: 0, by: '', time: 0 };

const ComponentsArray: Array<{
	component: React.FC<unknown> | React.ComponentClass;
	attrs: Record<string, unknown>;
}> = [
	{ component: App, attrs: {} },
	{
		component: Comment,
		attrs: {
			comment: testComment,
			depth: 0,
			toggleCollapse: () => {
				null;
			},
			collapsed: false,
		},
	},
	{ component: CommentTree, attrs: { parent: testComment, depth: 0 } },
	{ component: ItemList, attrs: { items: [testComment] } },
	{ component: ItemListSingle, attrs: { item: testComment } },
	{ component: ItemMeta, attrs: { by: '', time: 0, id: 0 } },
	{ component: ItemsPage, attrs: { type: 'top' } },
	{ component: Loading, attrs: {} },
	{ component: Nav, attrs: {} },
	{
		component: PostPage,
		attrs: {
			location: { state: { item: testComment } },
		},
	},
	{ component: UserPage, attrs: { location: { search: '?id=abc' } } },
];

type NJSGlobal = NodeJS.Global;

interface CustomNodeJsGlobal extends NJSGlobal {
	fetch: unknown;
}

declare const global: CustomNodeJsGlobal;

describe('Snapshots are matched for ', () => {
	global.fetch = jest.fn(() => Promise.resolve());
	ComponentsArray.forEach((ComponentObject) => {
		it(`${ComponentObject.component.name} component`, () => {
			const { component: Comp, attrs } = ComponentObject;
			const { asFragment } = render(
				<Router>
					<Comp {...attrs} />
				</Router>
			);
			expect(asFragment()).toMatchSnapshot();
		});
	});
});
