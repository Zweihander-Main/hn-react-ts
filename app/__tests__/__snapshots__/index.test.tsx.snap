// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshots are matched for  App component 1`] = `
<DocumentFragment>
  <div
    class="light"
  >
    <div
      class="container"
    >
      <nav
        class="row space-between"
      >
        <ul
          class="row nav"
        >
          <li>
            <a
              aria-current="page"
              class="nav-link active"
              href="/"
              style="color: rgb(187, 46, 31);"
            >
              Top
            </a>
          </li>
          <li>
            <a
              class="nav-link"
              href="/new"
            >
              New
            </a>
          </li>
          <li>
            <a
              class="nav-link"
              href="/best"
            >
              Best
            </a>
          </li>
          <li>
            <a
              class="nav-link"
              href="/ask"
            >
              Ask
            </a>
          </li>
          <li>
            <a
              class="nav-link"
              href="/show"
            >
              Show
            </a>
          </li>
          <li>
            <a
              class="nav-link"
              href="/job"
            >
              Job
            </a>
          </li>
        </ul>
        <button
          class="btn-clear"
          style="font-size: 30px;"
        >
          ðŸ”¦
        </button>
      </nav>
      <p
        style="font-size: 35px; margin-top: 20px; text-align: center;"
      >
        Loading
      </p>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Snapshots are matched for  Comment component 1`] = `
<DocumentFragment>
  <div
    class="comment"
    style="margin-left: 0px;"
  >
    <a
      class="toggleCollapse toggle-light"
      href="#"
    >
      [-]
    </a>
    <div
      class="meta-info-light"
    >
      <span>
        by 
        <a
          href="/user?id="
        />
      </span>
      <span>
        on 12/31/1969, 7:00 PM
      </span>
    </div>
    <p />
  </div>
</DocumentFragment>
`;

exports[`Snapshots are matched for  CommentTree component 1`] = `
<DocumentFragment>
  <p
    style="font-size: 35px; margin-top: 20px; text-align: center;"
  >
    Loading comment
  </p>
</DocumentFragment>
`;

exports[`Snapshots are matched for  ItemList component 1`] = `
<DocumentFragment>
  <ul>
    <div
      class="infinite-scroll-component__outerdiv"
    >
      <div
        class="infinite-scroll-component "
        style="height: auto; overflow: auto;"
      >
        <li
          class="post"
        >
          <a
            class="link"
          />
          <div
            class="meta-info-light"
          >
            <span>
              by 
              <a
                href="/user?id="
              />
            </span>
            <span>
              on 12/31/1969, 7:00 PM
            </span>
          </div>
        </li>
      </div>
    </div>
  </ul>
</DocumentFragment>
`;

exports[`Snapshots are matched for  ItemListSingle component 1`] = `
<DocumentFragment>
  <li
    class="post"
  >
    <a
      class="link"
    />
    <div
      class="meta-info-light"
    >
      <span>
        by 
        <a
          href="/user?id="
        />
      </span>
      <span>
        on 12/31/1969, 7:00 PM
      </span>
    </div>
  </li>
</DocumentFragment>
`;

exports[`Snapshots are matched for  ItemsPage component 1`] = `
<DocumentFragment>
  <p
    style="font-size: 35px; margin-top: 20px; text-align: center;"
  >
    Loading
  </p>
</DocumentFragment>
`;

exports[`Snapshots are matched for  Loading component 1`] = `
<DocumentFragment>
  <p
    style="font-size: 35px; margin-top: 20px; text-align: center;"
  >
    Loading
  </p>
</DocumentFragment>
`;

exports[`Snapshots are matched for  Nav component 1`] = `
<DocumentFragment>
  <nav
    class="row space-between"
  >
    <ul
      class="row nav"
    >
      <li>
        <a
          aria-current="page"
          class="nav-link active"
          href="/"
          style="color: rgb(187, 46, 31);"
        >
          Top
        </a>
      </li>
      <li>
        <a
          class="nav-link"
          href="/new"
        >
          New
        </a>
      </li>
      <li>
        <a
          class="nav-link"
          href="/best"
        >
          Best
        </a>
      </li>
      <li>
        <a
          class="nav-link"
          href="/ask"
        >
          Ask
        </a>
      </li>
      <li>
        <a
          class="nav-link"
          href="/show"
        >
          Show
        </a>
      </li>
      <li>
        <a
          class="nav-link"
          href="/job"
        >
          Job
        </a>
      </li>
    </ul>
    <button
      class="btn-clear"
      style="font-size: 30px;"
    >
      ðŸ”¦
    </button>
  </nav>
</DocumentFragment>
`;

exports[`Snapshots are matched for  PostMetaInfo component 1`] = `
<DocumentFragment>
  <div
    class="meta-info-light"
  >
    <span>
      by 
      <a
        href="/user?id="
      />
    </span>
    <span>
      on 12/31/1969, 7:00 PM
    </span>
  </div>
</DocumentFragment>
`;

exports[`Snapshots are matched for  PostPage component 1`] = `
<DocumentFragment>
  <p
    style="font-size: 35px; margin-top: 20px; text-align: center;"
  >
    Fetching post
  </p>
</DocumentFragment>
`;

exports[`Snapshots are matched for  UserPage component 1`] = `
<DocumentFragment>
  <p
    style="font-size: 35px; margin-top: 20px; text-align: center;"
  >
    Fetching User
  </p>
</DocumentFragment>
`;
